tests/unit/core_tests.py

```diff
@@ -341,3 +346,7 @@ def test_from_upstream_handles_no_sampled_or_flags(self):
         span = TraceInfo.from_upstream(1, 2, 3, None, None)
         self.assertIsNone(span.sampled)
         self.assertIsNone(span.flags)
+
+    def test_from_upstream_handles_no_parent(self):
+        span = TraceInfo.from_upstream(1, None, 3, None, None)
+        self.assertIsNone(span.parent_id)
```

Comment:
Just a suggestion: it might be nice to include a bit of context here with a brief comment.

In my experience this is the kind of test that people come back to later on and think "why are we testing this? can this even happen in real code? of course there's always a parent span because of [xyz reason that doesn't apply in your case]" and end up breaking things. Since this is testing a non-obvious but important scenario, you could add a line like

```python
def test_from_upstream_handles_no_parent(self):
    # Ensure that TraceInfo can be constructed with no parent span. This can happen when
    # the root span is created by this service.
    ...
```